x-pack/legacy/plugins/lens/server/routes/index_stats.ts

```diff
@@ -0,0 +1,177 @@
+  const router = setup.http.createRouter();
+  router.post(
+    {
+      path: '/index_stats/{indexPatternTitle}',
+      validate: {
+        params: schema.object({
+          indexPatternTitle: schema.string(),
+        }),
+        body: schema.object({
+          earliest: schema.string(),
+          latest: schema.string(),
+          timeZone: schema.maybe(schema.string()),
+          timeFieldName: schema.string(),
+          size: schema.number(),
+          fields: schema.arrayOf(
+            schema.object({
+              name: schema.string(),
+              type: schema.string(),
+            })
+          ),
+        }),
+      },
+    },
+    async (context, req, res) => {
+      const requestClient = context.core.elasticsearch.dataClient;
+
+      const indexPatternsService = new IndexPatternsService(requestClient.callAsCurrentUser);
+
+      const { earliest, latest, timeZone, timeFieldName, fields, size } = req.body;
+
+      try {
+        const indexPattern = await indexPatternsService.getFieldsForWildcard({
+          pattern: req.params.indexPatternTitle,
+          // TODO: Pull this from kibana advanced settings
+          metaFields: ['_source', '_id', '_type', '_index', '_score'],
+        });
+
+        const results = (await requestClient.callAsCurrentUser('search', {
+          index: req.params.indexPatternTitle,
+          body: {
+            query: {
+              bool: {
+                filter: [
+                  {
+                    range: {
+                      [timeFieldName]: {
+                        gte: earliest,
+                        lte: latest,
+                        time_zone: timeZone,
+                      },
+                    },
+                  },
+                ],
+              },
+            },
+            size,
+          },
+        })) as SearchResponse<Document>;
+
+        if (results.hits.hits.length) {
+          return res.ok({
+            body: recursiveFlatten(results.hits.hits, indexPattern, fields),
+          });
+        }
+        return res.ok({ body: {} });
+      } catch (e) {
+        if (e.isBoom) {
+          return res.internalError(e);
+        } else {
+          return res.internalError({
+            body: Boom.internal(e.message || e.name),
+          });
+        }
+      }
+    }
+  );
+}
+
+export function recursiveFlatten(
+  docs: Array<{
+    _source: Document;
+  }>,
+  indexPattern: FieldDescriptor[],
+  fields: Fields
+): Record<
+  string,
+  {
+    count: number;
+    cardinality: number;
+  }
+> {
+  const overallKeys: Record<
+    string,
+    {
+      count: number;
+      samples: unknown[];
+    }
+  > = {};
+
+  const expectedKeys: Record<string, boolean> = {};
+  fields.forEach(field => {
+    expectedKeys[field.name] = true;
+  });
+
+  // TODO: Alias types
+  indexPattern.forEach(field => {
+    if (!expectedKeys[field.name]) {
+      return;
+    }
+
+    let matches;
```

Comment:
I think this could be simplified to:

```js
const matches = docs.map(doc => {
  if (!doc) {
    return;
  }

  return get(doc._source, field.parent || field.name);
});
```

In fact this entire block could probably be simplified:

```js
indexPattern.forEach(field => {
  if (!expectedKeys.has(field.name)) {
    return;
  }

  docs.forEach(doc => {
    if (!doc) {
      return;
    }

    const match = get(doc._source, field.parent || field.name);
    const record = overallKeys[field.name];
    if (record) {
      record.count += 1;
      record.samples.add(match);
    } else {
      overallKeys[field.name] = {
        count: 1,
        // Using a set here makes the most sense and avoids the later uniq computation
        samples: new Set(match),
      };
    }
  });
});

return _.mapValues(overallKeys, record => ({
  count: record.count,
  cardinality: record.samples.size,
}));
```

Reply:
Good idea using a Set- the only change I had to make here is to the return type, which I had set up as `Record<string, { count, cardinality }>` and is an array here.